### PR TITLE
JAMES-2711 DMVRoutes Delete Api

### PR DESCRIPTION
**Pull request by hoangdat**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    DeletedMessagesVaultDeleteTask(DeletedMessageVault vault, User user, MessageId messageId) {
+        this.vault = vault;
+        this.user = user;
+        this.messageId = messageId;
+    }
+
+    @Override
+    public Result run() {
+        return Mono.from(vault.delete(user, messageId))
+            .doOnError(e -> LOGGER.error("Error while deleting message in DeletedMessageVault: {}", e.getMessage()))
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java]: Please don't loose the stacktrace

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public String type() {
+        return TYPE;
+    }
+
+    @Override
+    public Optional<TaskExecutionDetails.AdditionalInformation> details() {
+        return Optional.empty();
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java]: We need additional information about user, messageId

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java [hunk 3]
```diff
@@ -2100,6 +2106,138 @@ void purgeShouldProduceAFailedTaskWithVaultDeleteError() {
...
+                .get(taskId + "/await")
+            .then()
+                .body("status", is("completed"))
+                .body("taskId", is(taskId))
+                .body("type", is(DeletedMessagesVaultDeleteTask.TYPE))
+                .body("startedDate", is(notNullValue()))
+                .body("submitDate", is(notNullValue()))
+                .body("completedDate", is(notNullValue()));
+        }
+
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: We don't check that:
 - delete does delete the message in the vault?
 - delete does return a task id, with the right location header?

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java (hunk 1 above)
Comment (trantienduchn) [reply server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java]: you can log information about `user` & `messageId` here

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java [hunk 4]
```diff
@@ -212,6 +221,40 @@ private TaskIdDto globalActions(Request request, Response response) {
...
+            required = true,
+            name = "messageId",
+            paramType = "path parameter",
+            dataType = "String",
+            defaultValue = "none",
+            value = "Compulsory, Need to be a valid messageId")
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - user param is invalid"),
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java]: invalid `messageId` also returns 400 response

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java [hunk 5]
```diff
@@ -2100,6 +2106,138 @@ void purgeShouldProduceAFailedTaskWithVaultDeleteError() {
         }
     }
 
+    @Nested
+    class DeleteTest {
+
+        @Test
+        void deleteShouldProduceASuccessfulTaskEvenNoDeletedMessageExisted() {
+            String taskId =
+                when()
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: `with()`

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java [hunk 6]
```diff
@@ -2100,6 +2106,138 @@ void purgeShouldProduceAFailedTaskWithVaultDeleteError() {
...
+                .body("completedDate", is(notNullValue()));
+        }
+
+        @Test
+        void deleteShouldProduceASuccessfulTask() {
+            vault.append(USER, DELETED_MESSAGE, new ByteArrayInputStream(CONTENT)).block();
+            vault.append(USER, DELETED_MESSAGE_2, new ByteArrayInputStream(CONTENT)).block();
+
+            String taskId =
+                when()
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: idem

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java [hunk 7]
```diff
@@ -2100,6 +2106,138 @@ void purgeShouldProduceAFailedTaskWithVaultDeleteError() {
...
+                .body("completedDate", is(notNullValue()));
+        }
+
+        @Test
+        void deleteShouldNotAppendMessagesToUserMailbox() throws Exception {
+            vault.append(USER, DELETED_MESSAGE, new ByteArrayInputStream(CONTENT)).block();
+            vault.append(USER, DELETED_MESSAGE_2, new ByteArrayInputStream(CONTENT)).block();
+
+            String taskId =
+                when()
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: idem

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java [hunk 8]
```diff
@@ -2100,6 +2106,138 @@ void purgeShouldProduceAFailedTaskWithVaultDeleteError() {
...
+        class FailingDeleteTest {
+
+            @Test
+            void deleteShouldProduceAFailedTask() {
+                doReturn(Mono.error(new RuntimeException("mock exception")))
+                    .when(vault)
+                    .delete(any(), any());
+
+                String taskId =
+                    when()
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: idem

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java [hunk 9]
```diff
@@ -109,7 +111,10 @@ public String getValue() {
     public static final String USERS = "users";
     public static final String USER_PATH = ROOT_PATH + SEPARATOR + USERS;
     private static final String USER_PATH_PARAM = "user";
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java]: Why not using the same pattern as `MESSAGE_ID_PARAM` variable, as : 
`USER_PATH_PARAM = ":user"` instead?

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java (hunk 1 above)
Comment (trantienduchn) [reply server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultDeleteTask.java]:
```
LOGGER.error("Error while deleting message {} of user {} in DeletedMessageVault", messageId.serialize(), user.asString(), e)
```

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java (hunk 3 above)
Comment (hoangdat) [reply server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: > delete does return a task id, with the right location header?

I can not get your point for this test case.

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java]: See exportShouldReturnATaskCreated

**Comment by chibenwa** [comment]
We still miss the MD doc

File: src/site/markdown/server/manage-webadmin.md [hunk 10]
```diff
@@ -2841,6 +2842,35 @@ You need to create a CRON job to call this endpoint in a regular basis. Example:
...
+ - 400: Bad request: 
+   - user parameter is invalid
+   - messageId parameter is invalid
+ - 404: User not found
+ 
+ The scheduled task will have the following type `deletedMessages/delete` and the following `additionalInformation`:
+ 
+ ```
+ {
+   "userDeleteTo": "userDeleteTo@domain.ext",
```
Comment (chibenwa) [inline src/site/markdown/server/manage-webadmin.md]: Should be `"user"` simply, no?

**Comment by trantienduchn** [comment]
```
[61bc97416d83335a9fa552663cf7e587defc4333] [ERROR] Errors: 
[61bc97416d83335a9fa552663cf7e587defc4333] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:628 » ConditionTimeout
```
test this please

**Comment by hoangdat** [comment]
```
[61bc97416d83335a9fa552663cf7e587defc4333] [ERROR] Failures: 
[61bc97416d83335a9fa552663cf7e587defc4333] [ERROR]   WebAdminServerIntegrationTest.postShouldDoMigrationAndUpdateToTheLatestVersion:291 1 expectation failed.
[61bc97416d83335a9fa552663cf7e587defc4333] Response body doesn't match expectation.
[61bc97416d83335a9fa552663cf7e587defc4333] Expected: is "{\"version\":7}"
[61bc97416d83335a9fa552663cf7e587defc4333]   Actual: {"version":6}
```
test this please

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java [hunk 11]
```diff
@@ -173,14 +192,78 @@ public void define(Service service) {
...
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - action is invalid"),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side.")
+    })
+    private TaskIdDto globalActions(Request request, Response response) {
+        Task vaultTask = generateVaultTask(request);
+        TaskId taskId = taskManager.submit(vaultTask);
+        return TaskIdDto.respond(response, taskId);
+    }
+
+    @DELETE
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java]: Should be POST... 

 IMO `DELETE` returns 201

Here we create a related REST resource (a task) so a POST seems appropriate. The fact that this tasks actually DELETE the original endoint resources is somehow a side effect that might (hopefully) arise in the future resulting from this task execution.

IMO let's look at some work of people better than us at REST https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html : DeleteByQuery is a POST in ElasticSearch.

Agree with @rouazana on this: **it should be a POST**

**Comment by hoangdat** [comment]
Combined in #2296. Closed

**Pull request closed without merging**